src/framework/auth/strategies/oauth2/oauth2-strategy.ts

```diff
@@ -199,13 +199,17 @@ export class NbOAuth2AuthStrategy extends NbAuthStrategy {
     return this.http.post(url, this.buildRefreshRequestData(token), this.buildAuthHeader())
       .pipe(
         map((res) => {
+          const refreshedToken = <NbAuthOAuth2Token>this.createToken(res);
```

Comment:
let's just move this into a new method:

```
createRefreshedToken(res): NbAuthRefreshableToken {
...
}
```
and then call it almost as before:
```
return new NbAuthResult(
            true,
            res,
            this.getOption('redirect.success'),
            [],
            this.getOption('defaultMessages'),
            this.createRefreshedToken(res));
```

Reply:
@nnixaa 
I think you mean 
```
createRefreshedToken(res, existingToken): NbAuthOAuth2Token {
...
}
```
since we need the existing token for extracting the existing refresh_token

And NbAuthOAuth2Token instead of NbAuthRefreshableToken so that we return something assignable to NbAuthResult.token i.e NbAuthToken

Reply:
@alain-charles yes, of course, you are right.
But in the second part - return `NbAuthOAuth2Token` - is this always true? as by design we can assign any token to any strategy, so it is not necessary `NbAuthOAuth2Token`, but rather `NbAuthToken`.

Reply:
@nnixaa here is the best implementation i have found
```
  protected createRefreshedToken(res, existingToken: NbAuthRefreshableToken): NbAuthToken {
    const refreshedToken = this.createToken(res);
    if (isNbAuthRefreshableToken(refreshedToken)
        && !refreshedToken.getRefreshToken()
        && existingToken.getRefreshToken()) {
      refreshedToken.setRefreshToken(existingToken.getRefreshToken());
    }
    return refreshedToken;
  }
```

i think this is the one that is the most in the spirit of your architecture